### PR TITLE
markdown: improve preprocessors type

### PR DESCRIPTION
**Pull request by hauntsaninja**
https://github.com/Python-Markdown/markdown/blob/b701c34ebd7b2d0eb319517b9a275ddf0c89608d/markdown/preprocessors.py#L46

### CODE REVIEW

File: third_party/2and3/markdown/preprocessors.pyi [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 
 def build_preprocessors(md, **kwargs): ...
 
 class Preprocessor(util.Processor):
-    def run(self, lines) -> None: ...
+    def run(self, lines: List[str]) -> List[str]: ...
 
 class NormalizeWhitespace(Preprocessor):
-    def run(self, lines): ...
+    def run(self, lines: List[str]) -> List[str]: ...
```
Comment (srittau) [inline third_party/2and3/markdown/preprocessors.pyi:11]: While having a tight type in the (loosely) abstract `Preprocessor` class makes sense, I suggest to use `Iterable` here in the implementation. Same below.

File: third_party/2and3/markdown/preprocessors.pyi [hunk 2]
```diff
@@ -17,10 +17,10 @@ class HtmlBlockPreprocessor(Preprocessor):
     attrs_re: Any
     left_tag_re: Any
     markdown_in_raw: bool = ...
-    def run(self, lines): ...
+    def run(self, lines: List[str]) -> List[str]: ...
 
 class ReferencePreprocessor(Preprocessor):
```
Comment (srittau) [inline third_party/2and3/markdown/preprocessors.pyi:22]: This class doesn't exist in the file you linked to?

**Review by srittau — CHANGES_REQUESTED** [review]

File: third_party/2and3/markdown/preprocessors.pyi (hunk 2 above)
Comment (hauntsaninja) [reply third_party/2and3/markdown/preprocessors.pyi:22]: Good spot, looks like it got removed eight days ago, changes haven't been released

File: third_party/2and3/markdown/preprocessors.pyi (hunk 1 above)
Comment (hauntsaninja) [reply third_party/2and3/markdown/preprocessors.pyi:11]: str.split returns a List, so I believe this is the best type.

Comment (srittau) [reply third_party/2and3/markdown/preprocessors.pyi:11]: I meant the argument type, sorry for being ambiguous.

File: third_party/2and3/markdown/preprocessors.pyi (hunk 2 above)
Comment (srittau) [reply third_party/2and3/markdown/preprocessors.pyi:22]: Good timing. Out of scope for this PR then.

File: third_party/2and3/markdown/preprocessors.pyi (hunk 1 above)
Comment (hauntsaninja) [reply third_party/2and3/markdown/preprocessors.pyi:11]: Oh right, that makes way more sense

File: third_party/2and3/markdown/preprocessors.pyi [hunk 3]
```diff
@@ -1,14 +1,14 @@
-from typing import Any, Pattern
+from typing import Any, List, Pattern
```
Comment (hauntsaninja) [inline third_party/2and3/markdown/preprocessors.pyi:1]:
```suggestion
from typing import Any, Iterable, List, Pattern
```

File: third_party/2and3/markdown/preprocessors.pyi (hunk 1 above)
Comment (hauntsaninja) [reply third_party/2and3/markdown/preprocessors.pyi:11]:
```suggestion
    def run(self, lines: Iterable[str]) -> List[str]: ...
```

File: third_party/2and3/markdown/preprocessors.pyi [hunk 4]
```diff
@@ -17,10 +17,10 @@ class HtmlBlockPreprocessor(Preprocessor):
     attrs_re: Any
     left_tag_re: Any
     markdown_in_raw: bool = ...
-    def run(self, lines): ...
+    def run(self, lines: List[str]) -> List[str]: ...
```
Comment (hauntsaninja) [inline third_party/2and3/markdown/preprocessors.pyi:20]:
```suggestion
    def run(self, lines: Iterable[str]) -> List[str]: ...
```

File: third_party/2and3/markdown/preprocessors.pyi (hunk 2 above)
Comment (hauntsaninja) [reply third_party/2and3/markdown/preprocessors.pyi:22]: For what it's worth it looks like this one actually did need a List

**Review by srittau — APPROVED** [review]

**Pull request merged**